faiss/impl/ProductAdditiveQuantizer.h

```diff
@@ -0,0 +1,128 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <vector>
+
+#include <faiss/Index.h>
+#include <faiss/impl/AdditiveQuantizer.h>
+#include <faiss/impl/LocalSearchQuantizer.h>
+#include <faiss/impl/ResidualQuantizer.h>
+
+namespace faiss {
+
+/** Product Additive Quantizers
+ * TODO: add some docs
+ */
+struct ProductAdditiveQuantizer : AdditiveQuantizer {
```

Comment:
why is this a sub-class of additivequantizer ? If you override the decoding method, then there is little left from AQ no?

Reply:
I make it a sub-class of AQ because the search code of AQs could be re-used in the future (e.g. IndexAdditiveQuantizer and IndexIVFAdditiveQuantizer ).